core/partitioning/partitioning.cpp

```diff
@@ -0,0 +1,232 @@
+              seg_block.registerOutput(node_output);
+          }
+          if (!seg_block.raw_outputs().empty())
+            break;
+        }
+      }
+    }
+  }
+  // erase segments which still have no output
+  segmented_blocks.erase(
+      std::remove_if(
+          segmented_blocks.begin(),
+          segmented_blocks.end(),
+          [](SegmentedBlock& seg_block) { return seg_block.raw_outputs().empty(); }),
+      segmented_blocks.end());
+
+  return;
+}
+
+std::vector<SegmentedBlock> segment_graph(std::shared_ptr<torch::jit::Graph> g, const PartitionInfo& partition_info) {
+  auto min_block_size = partition_info.min_block_size;
+  std::unordered_set<std::string> forced_fallback_operators(
+      partition_info.forced_fallback_operators.begin(), partition_info.forced_fallback_operators.end());
+
+  auto nodes = g->block()->nodes();
+  std::vector<SegmentedBlock> segmented_blocks;
+
+  // segment the nodes
+  std::vector<torch::jit::Node*> tensorrt_nodes, pytorch_nodes;
+  for (const auto n : nodes) {
+    if (n->kind() == torch::jit::prim::Constant)
+      continue;
+
+    std::string node_string(n->kind().toQualString());
+    if (conversion::OpSupported(n) && !forced_fallback_operators.count(node_string)) {
+      tensorrt_nodes.push_back(n);
+      if (tensorrt_nodes.size() >= min_block_size && !pytorch_nodes.empty()) {
+        segmented_blocks.emplace_back(SegmentedBlock::kTorch, pytorch_nodes);
+        pytorch_nodes.clear();
+      }
+    } else {
+      if (tensorrt_nodes.size() >= min_block_size) {
+        segmented_blocks.emplace_back(SegmentedBlock::kTensorRT, tensorrt_nodes);
+      } else {
+        pytorch_nodes.insert(pytorch_nodes.end(), tensorrt_nodes.begin(), tensorrt_nodes.end());
+      }
+      tensorrt_nodes.clear();
+      pytorch_nodes.push_back(n);
+    }
+  }
+
+  // if there is any kTorch nodes left, then either the last nodes are kTorch or last nodes are kTensorRT but num <
+  // min_block_size
+  if (!pytorch_nodes.empty()) {
+    pytorch_nodes.insert(pytorch_nodes.end(), tensorrt_nodes.begin(), tensorrt_nodes.end());
+    segmented_blocks.emplace_back(SegmentedBlock::kTorch, pytorch_nodes);
+  } else {
+    segmented_blocks.emplace_back(SegmentedBlock::kTensorRT, tensorrt_nodes);
+  }
+
+  return std::move(segmented_blocks);
+}
+
+std::vector<SegmentedBlock> Partition(
+    std::shared_ptr<torch::jit::Graph> g,
+    std::vector<ir::InputRange>& input_ranges,
+    const PartitionInfo& partition_info) {
+  LOG_DEBUG(partition_info);
+  // segment lowering global graph into blocks
+  std::vector<SegmentedBlock> segmented_blocks = segment_graph(g, partition_info);
+
+  // resolve nonTensor inputs/outputs
+  resolveNonTensorInputs(segmented_blocks, g);
+
+  // register input/output torch::jit::Value for segmented graphs
+  registerSegmentsOutputs(segmented_blocks, g);
+
+  // store the mapping from lowering graph torch::jit::Value => torch::jit::IValue that we get by running segments

```

Comment:
Can this be isolated in the shape analysis component or is it necessary else where?

Reply:
We can move it into shape analysis. But this function runs on all SegmentedBlock while shape analysis solely runs on each SegmentedBlock, we may need to refactor shape analysis as well.